### PR TITLE
Build render function with AST

### LINKED ISSUE #30: Precompile SFC render functions using AST
Relates to #29

I put together a little example to demonstrate how to construct render functions using the AST. This would allow you to parse the CGX template, and construct an optimal render function during the conversion, instead of at runtime.

```python
"""
Example of how to construct a function with the AST, compile it and
get a reference to it that you can actually call and hook up to a class.

Useful docs:
- https://docs.python.org/3/library/ast.html
- https://greentreesnakes.readthedocs.io/
"""

import ast


def make_attr_read(name, attr):
    return ast.Attribute(
        value=ast.Name(id=name, ctx=ast.Load()),
        attr=attr,
        ctx=ast.Load(),
    )


def make_function(name, args):
    arguments = ast.arguments(
        args=[ast.arg(a) for a in args],
        posonlyargs=[],
        kwonlyargs=[],
        kw_defaults=[],
        defaults=[],
    )
    return ast.FunctionDef(
        name=name,
        body=[],
        args=arguments,
        decorator_list=[],
    )


def make_module(body):
    return ast.Module(body=body, type_ignores=[])


def compile_tree(tree, name):
    ast.fix_missing_locations(tree)
    code = compile(tree, filename="<ast>", mode="exec")
    _globals = globals().copy()
    exec(code, _globals)
    return _globals[name]


# we'll recreate this tree by crafting it manually with ast nodes
tree = ast.parse(
    "def render(self):\n"
    "    return self.foo\n"
)


# now manually with nodes...
fn = make_function("render", ["self"])
fn.body = [
    ast.Return(value=make_attr_read("self", "foo")),
]
tree2 = make_module([fn])


# our "self"
class Foo:
    def __init__(self):
        self.foo = "Hello, world!"


obj = Foo()


if __name__ == "__main__":
    # let's test. you should see "Hello, world!" twice when running this
    fn = compile_tree(tree, "render")
    print(fn(obj))

    fn2 = compile_tree(tree2, "render")
    print(fn2(obj))
```

Note: I tried inserting breakpoints, and it works, but not too smoothly...

### PR DESCRIPTION
**Pull request by berendkleinhaneveld**
Closes #30 .

Create the render function of a component specified in a `.cgx` file with the help of an AST.

- [x] Check that the returned class definition is subclass of Component
- [x] Add syntactic sugar for `self.props` and `self.state`

### CODE REVIEW

File: collagraph/cgx/cgx.py [hunk 1]
```diff
@@ -54,214 +245,230 @@ class Foo(cg.Component):
...
+                    raise NameError(f"name '{name}' is not defined")
+                """
+            ),
+            mode="exec",
+        ).body[0],
+    )
 
-    if not component_type:
-        raise ValueError(f"No subclass of Component found:\n{script}")
+    # Find ClassDef
```
Comment (Korijn) [inline collagraph/cgx/cgx.py:283]: ultimately I guess you want to make this just as smart as the code you had behavior, to identify a subclass of Component, and not just any class definition

File: collagraph/components/__init__.py [hunk 2]
```diff
@@ -3,23 +3,7 @@
...
-
-    def __new__(cls, name, bases, attrs):
-        if cls.RENDER_FUNCTION:
-            attrs["render"] = cls.RENDER_FUNCTION
-
-        return type.__new__(cls, name, bases, attrs)
-
-
-class Component(metaclass=ComponentMeta):
+class Component:
```
Comment (Korijn) [inline collagraph/components/__init__.py:6]: I have to say it's pretty neat that this metaclass also vanishes 🚀

File: collagraph/cgx/cgx.py [hunk 3]
```diff
@@ -54,218 +54,339 @@ class Foo(cg.Component):
...
+    # Compile the tree into a code object (module)
+    code = compile(script_tree, filename="<ast>", mode="exec")
+    # Execute the code as module and pass a dictionary that will capture
+    # the global and local scope of the module
+    module_namespace = {}
+    exec(code, module_namespace)
+
+    # Check that the class definition is an actual subclass of Component
+    component_class = module_namespace[component_def.name]
+    if not issubclass(component_class, Component):
```
Comment (Korijn) [inline collagraph/cgx/cgx.py:144]: seems like it's pretty easy to trigger this if you define e.g. a dataclass or something else in your cgx file, before your component class definition

Comment (berendkleinhaneveld) [reply collagraph/cgx/cgx.py:144]: Yes. Without evaluating/executing the SFC module, it's not possible to know whether a ClassDef is a sub(sub)class of Component, so I think we should just document that the first class definition in your SFC should be a component sub(sub)class.

Comment (Korijn) [reply collagraph/cgx/cgx.py:144]: Or... The last? I think that might be more practical, if you think about reasons to define more classes in a cgx file

Comment (berendkleinhaneveld) [reply collagraph/cgx/cgx.py:144]: 👍makes sense

Comment (berendkleinhaneveld) [reply collagraph/cgx/cgx.py:142]: I've also added a test to make this behaviour a bit more 'documented'.

**Review by Korijn — APPROVED** [review]

**Pull request merged**